### PR TITLE
V kalc/tab adaptive cards

### PR DESCRIPTION
**Pull request by v-kalc**
Sample code to display adaptive cards in tab using Node js

### CODE REVIEW

File: samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/AdapterWithErrorHandler.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+//
```
Comment (v-chetsh) [inline samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/AdapterWithErrorHandler.cs:3]: Remove empty line

File: samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs [hunk 2]
```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+//
```
Comment (v-chetsh) [inline samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs]: Remove empty line

File: samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                    Tab = new TabResponsePayload
+                    {
+                        Type = "continue",
+                        Value = new TabResponseCards
+                        {
+                            Cards = new List<TabResponseCard>
+                            {
+                                new TabResponseCard
+                                {
+                                    Card = GetAdaptiveCard1()
```
Comment (v-chetsh) [inline samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs:43]: Please provider proper name.
Don't use 1 or 2

File: samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+                                },
+                                new TabResponseCard
+                                {
+                                    Card = GetAdaptiveCard2()
```
Comment (v-chetsh) [inline samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs:47]: Please provider proper name.
Don't use 1 or 2

File: samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+                    Tab = new TabResponsePayload
+                    {
+                        Type = "continue",
+                        Value = new TabResponseCards
+                        {
+                            Cards = new List<TabResponseCard>
+                            {
+                                new TabResponseCard
+                                {
+                                    Card = GetAdaptiveCard2()
```
Comment (v-chetsh) [inline samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs:67]: Please provider proper name.
Don't use 1 or 2

File: samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+                            },
+                        },
+                    },
+                });
+            }
+
+            return null;
+        }
+
+        private AdaptiveCard GetAdaptiveCard1()
```
Comment (v-chetsh) [inline samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs:82]: Add summary

File: samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+                        Text = "Test the tab view with Adaptive card",
+                        Weight = AdaptiveTextWeight.Bolder,
+                    },
+                },
+
+                Actions = new List<AdaptiveAction>
+                {
+                    new AdaptiveSubmitAction
+                    {
+                        Title = "Action 1",
```
Comment (v-chetsh) [inline samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs:99]: Add some meaningful action name based on the action to perform

File: samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+                        Spacing = AdaptiveSpacing.Medium,
+                    },
+                    new AdaptiveTextBlock
+                    {
+                        Text = "tab/submit request is triggered to your bot with the corresponding data through the Action.Submit function of Adaptive Card",
+                        Spacing = AdaptiveSpacing.Medium,
+                        Wrap = true,
+                    },
+                },
+
```
Comment (v-chetsh) [inline samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs:142]: Remove empty line

File: samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Controllers/BotController.cs [hunk 9]
```diff
@@ -0,0 +1,37 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+//
```
Comment (v-chetsh) [inline samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Controllers/BotController.cs:3]: Remove empty line

File: samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Startup.cs [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+                .UseStaticFiles()
+                .UseWebSockets()
+                .UseRouting()
+                .UseAuthorization()
+                .UseEndpoints(endpoints =>
+                {
+                    endpoints.MapControllers();
+                });
+
+            // app.UseHttpsRedirection();
```
Comment (v-chetsh) [inline samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Startup.cs:57]: Remove if not required

File: samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/AppPackage/manifest.json [hunk 11]
```diff
@@ -0,0 +1,45 @@
+﻿{
+  "$schema": "https://raw.githubusercontent.com/OfficeDev/microsoft-teams-app-schema/preview/DevPreview/MicrosoftTeams.schema.json",
+  "manifestVersion": "1.9",
+  "version": "1.0.0",
+  "id": "4d9e95ab-ebe3-4976-82e9-3526d5c54906",
+  "packageName": "com.microsoft.sequentialflowbot",
```
Comment (v-chetsh) [inline samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/AppPackage/manifest.json:6]: Please change package name according to your application.

File: samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/AppPackage/manifest.json [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+      "entityId": "homeTab",
+      "name": "Home",
+      "contentBotId": "{{BOT-ID}}",
+      "scopes": [ "personal" ]
+    },
+    {
+      "entityId": "moreTab",
+      "name": "More",
+      "contentBotId": "{{BOT-ID}}",
+      "scopes": [ "personal" ]
```
Comment (v-chetsh) [inline samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/AppPackage/manifest.json:37]: Remove if not required

File: samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/AppPackage/manifest.json [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    }
+  ],
+  "permissions": [ "identity", "messageTeamMembers" ],
+  "validDomains": [
+    "*.ngrok.io",
+    "token.botframework.com"
```
Comment (v-chetsh) [inline samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/AppPackage/manifest.json:43]: Add dummy for deployed app domain also.

File: samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+    /// Bot Activity handler class
+    /// </summary>
+    public class ActivityBot : TeamsActivityHandler
+    {
+        /// <summary>
+        /// Invoked when an invoke activity is received from the connector. Invoke activities can be used to communicate many different things.
+        /// </summary>
+        /// <param name="turnContext"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
```
Comment (v-chetsh) [inline samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs:27]: Add proper summary also.

File: samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+                }));
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Sample Adaptive card
+        /// </summary>
+        /// <returns></returns>
```
Comment (v-chetsh) [inline samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs:81]: Kindly add proper summary

File: samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+            {
+                Body = new List<AdaptiveElement>
+                {
+                    new AdaptiveTextBlock
+                    {
+                        Text = "Test the tab view with Adaptive card",
+                        Weight = AdaptiveTextWeight.Bolder,
+                    },
+                },
+
```
Comment (v-chetsh) [inline samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs:94]: Remove empty line

File: samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+            };
+
+            return card;
+        }
+
+        /// <summary>
+        /// Sample Adaptive card
+        /// </summary>
+        /// <returns></returns>
+        private AdaptiveCard GetAdaptiveCard2()
```
Comment (v-chetsh) [inline samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/Bots/ActivityBot.cs:115]: Please add proper summary & rename method name

**Review by v-chetsh — APPROVED** [review]
All looks good now.
Approving it.

File: samples/tab-adaptive-cards/nodejs/Readme.md [hunk 18]
```diff
@@ -0,0 +1,123 @@
+# Tab With AdpativeCard Flow
+
+This App talks about the Teams tab which displays Adaptive card with Node JS
```
Comment (Wajeed-msft) [inline samples/tab-adaptive-cards/nodejs/Readme.md:3]: Let's hyperlink to docs: https://docs.microsoft.com/en-us/microsoftteams/platform/tabs/how-to/build-adaptive-card-tabs

File: samples/tab-adaptive-cards/nodejs/Readme.md [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+    - **Zip** up the contents of the `appPackage` folder to create a `manifest.zip`
+    - **Upload** the `manifest.zip` to Teams (in the Apps view click "Upload a custom app")
+
+## Interacting with the bot in Teams
+
+You can interact with this bot by `@Sequential Workflows` (BotName). The bot will respond with adaptive card requesting you the details.
+
+- Install App
+
+Navigate to `Manage apps` > `Upload a custom app` (Bottom-Right of the screen) > Upload `manifest.zip` > `Add`
```
Comment (Wajeed-msft) [inline samples/tab-adaptive-cards/nodejs/Readme.md:97]: This looks like section for another sample. Could you please update the these steps for Tab?

File: samples/tab-adaptive-cards/nodejs/Readme.md [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+![Sign in Card](Images/sign-in.png)
+
+- Sign in popup
+
+![Sing in popup](Images/sign-in-popup.png)
+
+- Sign out card
+
+![Sign out Card](Images/sign-out.png)
+
```
Comment (Wajeed-msft) [inline samples/tab-adaptive-cards/nodejs/Readme.md:30]: Could you please aslo add screenshot of "More" Tab?

**Review by Wajeed-msft — CHANGES_REQUESTED** [review]

File: samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/AppPackage/manifest.json (hunk 11 above)
Comment (AbhijeetThakur-msft) [reply samples/tab-adaptive-cards/csharp/TabWithAdpativeCardFlow/AppPackage/manifest.json]: changed

File: samples/tab-adaptive-cards/nodejs/Readme.md (hunk 19 above)
Comment (AbhijeetThakur-msft) [reply samples/tab-adaptive-cards/nodejs/Readme.md:97]: Changed

File: samples/tab-adaptive-cards/nodejs/Readme.md (hunk 18 above)
Comment (AbhijeetThakur-msft) [reply samples/tab-adaptive-cards/nodejs/Readme.md:3]: added

File: samples/tab-adaptive-cards/nodejs/Readme.md (hunk 20 above)
Comment (AbhijeetThakur-msft) [reply samples/tab-adaptive-cards/nodejs/Readme.md:30]: removed this as this contains the same content

**Review by Wajeed-msft — APPROVED** [review]

**Pull request merged**